aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
     return at::native::nested_from_padded_generic(padded, sizes);
   }
 }
+
+Tensor _collapse_two_dims_3(Tensor input, int64_t dim1, int64_t dim2) {
+  TORCH_CHECK(dim1 > 0, "dim1: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 > 0, "dim2: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 - 1 == dim1, "dim2 must be one more than dim1.")
+  TORCH_CHECK(dim1 == 1, "dim1 must be 1.")
+  TORCH_CHECK(input.dim() == 3, "Expected input to be 3 dim.");
+
+  auto* nt_input = get_nested_tensor_impl(input);
+  TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+  Tensor nt_sizes = nt_input->get_nested_size_tensor();
+  const auto& input_buffer = nt_input->get_buffer();
+
+  Tensor sizes_dim1 = at::native::narrow(nt_sizes, 1, 0, 1);
+  Tensor sizes_dim2 = at::native::narrow(nt_sizes, 1, 1, 1);
+
+  Tensor new_nt_sizes;
+  if (dim1 == 1) {
+    Tensor collapsed_sizes = sizes_dim1 * sizes_dim2;
+    new_nt_sizes = collapsed_sizes.contiguous();
+  }
+  Tensor result = at::detail::make_tensor<NestedTensorImpl>(nt_input->get_buffer(), new_nt_sizes);
+  TORCH_CHECK(result.dim() == 2, "Expected result to be 2 dimensional.");
+  return result;
+}
+
+Tensor batch_offsets_from_efficient_size(Tensor ef_sizes) {
+  int64_t* nt_sizes_ptr = ef_sizes.data_ptr<int64_t>();
+  Tensor offsets = at::empty({1 + ef_sizes.size(0)}, at::kLong);
+  int64_t* offsets_ptr = offsets.data_ptr<int64_t>();
+  offsets_ptr[0] = 0;
+  int64_t ef_sizes_size_1 = ef_sizes.size(1);
+  for (int64_t i = 0; i < ef_sizes.size(0); i++) {
+    int64_t prod = 1;
+    for (int64_t j = 0; j < ef_sizes_size_1; j++) {
+      prod = prod * nt_sizes_ptr[i * ef_sizes_size_1 + j];
+    }
+    offsets_ptr[i + 1] = offsets_ptr[i] + prod;
+  }
+  return offsets;
+}
+
+Tensor NestedTensor_to_padded_tensor_cuda(const Tensor& t, double padding) {
+  if ((t.dim() >= 2 && t.dim() <= 4)) {
+    auto orig_nt_dim = t.dim();
+    auto* nt_input = get_nested_tensor_impl(t);
+    TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+    const auto& nt_buffer = nt_input->get_buffer();
+    const auto nt_input_opt_size_2 = nt_input->get_opt_size(2);
+
+    Tensor nt = t;
+    if (t.dim() == 3 && nt_input_opt_size_2) {
+      nt = _collapse_two_dims_3(nt, 1, 2);
+    }
+
+    nt_input = get_nested_tensor_impl(nt);
+    Tensor nt_sizes = nt_input->get_nested_size_tensor();
+    Tensor offsets = batch_offsets_from_efficient_size(nt_sizes);
+    auto new_size = NestedTensor_get_max_size(*nt_input);
+    new_size.insert(new_size.begin(), nt_sizes.size(0));
+    Tensor output = at::empty(IntArrayRef(new_size), nt_buffer.options());
+
+    int64_t input_dim = nt_sizes.size(1);
+    int64_t batch_size = nt_sizes.size(0);
+    at::Tensor metadata = at::cat({offsets, nt_sizes.reshape(-1)});
```

Comment:
we do some clever stuff in other paths to avoid the at::cat here; can we at least have a TODO to replicate it here?